### DISCUSSION TITLE
How to take out the important subgraphs of the whole graph?

### DISCUSSION THREAD

**Discussion by yung1231** [pyg-team/pytorch_geometric#5840] [Q&A]
Hello, I built a model to classify the whole graph and I tried to use `explainer.explain_graph` to get those parts that are important.

Then I got two returns `node_feat_mask`、`edge_mask`.

Then I use `explainer.visualize_subgraph` to visualize the result but I see that the output is still the whole graph, not the subgraph.

Is there a way to show only the important subgraphs instead of the whole graph? Because the number of nodes and edges of the whole graph is too large, which makes it impossible to observe.

---

Then there is one more question, each time I run the result, the values of `node_feat_mask` and `edge_mask` are different, is this normal?

---

```python
explainer = GNNExplainer(model, epochs=3)    # only be executed once

# The following execution, each time the result is different
node_feat_mask, edge_mask = explainer.explain_graph(data.x, data.edge_index) 
ax, G = explainer.visualize_subgraph(None, edge_index, edge_mask, y=data.y, threshold=0.5)  # mask less than
```

Thanks.

**Comment by rusty1s** [comment]
One solution would be to manually shrink your graph to only show important edges and nodes, e.g. via
`subgraph(node_feat_mask > 0.5, edge_index)`. You can visualize the result afterwards.

`GNNExplainer` has indeed some variance in its result, but you may be able to fine-tune [its hyperparameters](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/models/gnn_explainer.py#L62) a bit to get more stable results.

**Reply by yung1231** [reply]
But I don't have `node_mask`, only `edge_mask`.

Do I need to set `node_mask` manually or is there any other way? I can't set my node manually because I have thousands of nodes.

---

In `GNNExplainer`, I think there are only two parameters that can be adjusted, `epochs` and `lr`, but I have tried to adjust the parameters before asking the question, If I adjust the `epochs`, it is true that the final result will be a little better, but the final result will still change every time.

And then you mentioned the `coeffs` parameter, I'm not sure how to modify it.

**Reply by rusty1s** [reply]
It depends on what you want to do. If you simply want to mask out unimportant nodes, you can also do
```python
edge_index = edge_index[:, edge_mask > 0.5]
```

You can specify other hyperparameters in `GNNExplainer` during instantiation:
```python
GNNExplainer(..., edge_size=0.01)
```

**Reply by yung1231** [reply]
Sorry, I'm still not sure, I just want to show the important subgraphs now.

Then I use the function `subgraph(node_idx: int, x: Tensor, edge_index: Tensor, **kwargs)` as you mentioned above, using the function `subgraph(node_feat_mask > 0.5, edge_index)`.

But when I do this, I type `ret = explainer.subgraph(node_idx=node_feat_mask > 0.5, x=data.x, edge_index=edge_index)` and I get the following error

```python
IndexError                                Traceback (most recent call last)
/tmp/ipykernel_30671/1758357763.py in <module>
      3 # print(edge_index.shape)
      4 # edge_index = edge_index[:, edge_mask > 0.5]
----> 5 ret = explainer.subgraph(node_idx=node_feat_mask > 0.5, x=data.x, edge_index=edge_index)
      6 # print(ret)

~/env/lib/python3.7/site-packages/torch_geometric/nn/models/explainer.py in subgraph(self, node_idx, x, edge_index, **kwargs)
    211         subset, edge_index, mapping, edge_mask = k_hop_subgraph(
    212             node_idx, self.num_hops, edge_index, relabel_nodes=True,
--> 213             num_nodes=num_nodes, flow=self._flow())
    214 
    215         x = x[subset]

~/env/lib/python3.7/site-packages/torch_geometric/utils/subgraph.py in k_hop_subgraph(node_idx, num_hops, edge_index, relabel_nodes, num_nodes, flow)
    204     for _ in range(num_hops):
    205         node_mask.fill_(False)
--> 206         node_mask[subsets[-1]] = True
    207         torch.index_select(node_mask, 0, row, out=edge_mask)
    208         subsets.append(col[edge_mask])

IndexError: The shape of the mask [128] at index 0 does not match the shape of the indexed tensor [10] at index 0
```

One of my graphs has this information: `Data(x=[304, 128], edge_index=[2, 1014], y=0)`

> Each node is composed of a vector of 128 dimensions.

**Reply by rusty1s** [reply]
How does `node_feat_mask` looks like? It would need to be of type `feat_mask_type == "individual_feature"` or `"scalar"` to obtain a node-level mask from it. Otherwise, you do not know which node is important.

**Reply by yung1231** [reply]
So I explain the whole graph, if I want to show the important subgraph, do I have to go through each node one by one?

Like this

```python
explainer.subgraph(node_idx=0, x=data.x, edge_index=edge_index)

explainer.subgraph(node_idx=1, x=data.x, edge_index=edge_index)

explainer.subgraph(node_idx=2, x=data.x, edge_index=edge_index)

...
```

I don't know what to do to achieve my goal.

**Reply by rusty1s** [reply]
Yes, if you want to explain every single node, you have to run `GNNExplainer` for all of these.

**Reply by rusty1s** [reply]
[Here](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/gnn_explainer_ba_shapes.py) is an example.

**Reply by yung1231** [reply]
No no, what I want is to classify the whole graph(is graph classification, not node classification), and then only show the important subgraph (Instead of showing the whole graph, only the important subgraph is displayed).
I use `explainer.explain_graph` and then use  `explainer.visualize_subgraph` but the `explainer.visualize_subgraph` result shows the whole graph, not just the important subgraphs.
> I want to show only the important subgraphs because the number of nodes on the graph is so large that it is completely unobservable if the whole graph is displayed.

**Reply by rusty1s** [reply]
Yes, this brings us back to my comment earlier. You need to filter the "unimportant nodes" and edges on your own. For example, if `feat_mask_type = "scalar"`, you can do
```python
node_mask = node_feat_mask > 0.5
edge_mask = edge_mask > 0.5

edge_index = edge_index[:, edge_mask]  # Filter important edges
edge_index, _ = subgraph(node_mask, edge_index, relabel_nodes=True)  # Filter important nodes

**Reply by yung1231** [reply]
Ummm, It looks like something, but some use `edge_index, _ = subgraph(node_mask, edge_index)` to run the result, the `edge_index` will be empty, both large and small graphs have this situation.

**Reply by rusty1s** [reply]
You might want to adjust the thresholds to a lower value.
